src/dialog-user/services/dialogData.ts

```diff
@@ -171,6 +171,11 @@ export default class DialogDataService {
       }
     }
 
+    if (field.type === 'DialogFieldDateControl') {
+      validation.isValid = (fieldValue instanceof Date);
```

Comment:
I'm worried this will fail for dynamic fields - the server can't send a Date object, so we have to convert, and that's called only from `$onInit` it seems, not when updating a field

Reply:
I believe that `setupField` should take care of converting data to `Date` object, maybe I just need to include `DialogFieldDateTimeControl`, anyway I need to verify this works.

https://github.com/ManageIQ/ui-components/blob/0f871a032a7dfabf244240f99f4f02082f61127a/src/dialog-user/services/dialogData.ts#L38-L44

Reply:
looks like the Dialog User doesn't work well if the value is updated from Automate.

I've applied a simple Automate method (thanks @d-m-u):

```ruby
new_date = Time.utc(2019, 07, 02)
$evm.object['values'] = new_date
```

I'm not sure if the method is applied correctly, but from what happens I think the problem is in the Dialog User anyway:

1. I cleared out the field so it's empty -> the Submit button correctly gets disabled
2. I checked checkbox that should update the dynamic DialogFieldDateControl field
3. The field is updated with today's date (not `utc(2019, 07, 02)`)  -> the Submit button is still **disabled**

If I pick the date from calendar manually, the button gets **enabled**, as expected.

Now I'm using this to log the data to evm.log:

```ruby
trigger = $evm.root['dialog_trigger']

$evm.log(:info, "======= trigger value #{trigger}")
if trigger == "t"
  new_date = Time.utc(2019, 07, 02)
  $evm.object['values'] = new_date
  $evm.log(:info, "======== new date #{$evm.object['values']}")
else # "f"
  $evm.object['values'] = ''
  $evm.log(:info, "======== clearing up date #{$evm.object['values']}")
end

```
but can't get the log into evm.log nor automate.log, and in `setupField` I'm getting only `undefined`, so I might be using the Automate wrong as well.

Reply:
```diff
diff --git a/src/dialog-user/services/dialogData.ts b/src/dialog-user/services/dialogData.ts
index c8fe465..a8a0dd8 100644
--- a/src/dialog-user/services/dialogData.ts
+++ b/src/dialog-user/services/dialogData.ts
@@ -43,6 +43,14 @@ export default class DialogDataService {
       }
     }

+    if (field.type === 'DialogFieldDateControl') {
+      if (_.isNull(field.default_value) || _.isUndefined(field.default_value)) {
+        field.dateField = field.timeField = new Date();
+      } else {
+        field.dateField = field.timeField = new Date(data.default_value);
+      }
+    }
+
     field.default_value = this.setDefaultValue(field);

     return field;
@@ -95,14 +103,14 @@ export default class DialogDataService {
       defaultValue = data.values[firstOption][fieldValue];
     }

-    if (data.type === 'DialogFieldDateControl' || data.type === 'DialogFieldDateTimeControl') {
-      defaultValue = data.values ? new Date(data.values) : new Date();
-    }
-
     if (data.default_value) {
       defaultValue = data.default_value;
     }

+    if (data.type === 'DialogFieldDateControl' || data.type === 'DialogFieldDateTimeControl') {
+      defaultValue = data.dateField ? new Date(data.dateField) : new Date();
+    }
+
     if (data.type === 'DialogFieldDropDownList' && data.options.force_multi_value && data.default_value) {
       defaultValue = JSON.parse(data.default_value);
     }
```

With these changes I'm getting the values as `Date()` even after using Automate, only the validation doesn't get run.

I'm not receiving any values into `value` attribute, so using `default_value`.

There's a lot of confusion with using `values`, `value` or `default_value`. I'm not sure what exactly is correct, also I think DialogFieldDateTimePicker won't work correctly if it works with `value` instead of `values`.

Reply:
Yeah, it is pretty confusing. On the automate side, you want to set the `'value'` key like you're doing above:
`$evm.object['value'] = something`

However, on the ui-components side, for most of the field types, `default_value` is the value that is getting passed back from the refresh API call that we should be looking at to determine what to show to the user after a refresh happens. We changed it in https://github.com/ManageIQ/ui-components/commit/7dcb1f7ca9fc404588cfd8944796bd3ed726c376. For sorted items, `values` is the key we use since it needs to be a list, and `default_value` is simply the one that is selected from that list.

On the ui-components side, because of the way datetime controls work, there's special logic for the date and time parts because the `default_value` comes in as a string (cause it's just a JSON response), and then it gets parsed and separated into a `dateField` and a `timeField` since the controls are separate.